app/server/lib/OIDCConfig.ts

```diff
@@ -0,0 +1,192 @@
+ *    given_name
+ *    family_name
+ *
+ * Expected environment variables:
+ *    env GRIST_OIDC_SP_HOST=https://<your-domain>
+ *        Host at which our /oidc endpoint will live; identifies our application.
+ *    env GRIST_OIDC_IDP_ISSUER
+ *        The issuer URL for the IdP.
+ *    env GRIST_OIDC_IDP_CLIENT_ID
+ *        The client ID for the application, as registered with the IdP.
+ *    env GRIST_OIDC_IDP_CLIENT_SECRET
+ *        The client secret for the application, as registered with the IdP.
+ *    env GRIST_OIDC_IDP_SCOPES
+ *        The scopes to request from the IdP, as a space-separated list. Defaults to "openid email profile".
+ *
+ * This version of OIDCConfig has been tested with Keycloak OIDC IdP following the instructions
+ * at:
+ *   https://www.keycloak.org/getting-started/getting-started-docker
+ * When running on localhost and http, the settings tested were with:
+ *   - GRIST_OIDC_SP_HOST=http://localhost:8484 (or whatever port you use for Grist)
+ *   - GRIST_OIDC_IDP_ISSUER=http://localhost:8080/realms/myrealm (replace 8080 by the port you use for keycloak)
+ *   - GRIST_OIDC_IDP_CLIENT_ID=myclient
+ *   - GRIST_OIDC_IDP_CLIENT_SECRET=YOUR_SECRET (as set in keycloak)
+ *   - GRIST_OIDC_IDP_SCOPES="openid email profile"
+ */
+
+import * as express from 'express';
+import {GristLoginSystem, GristServer} from './GristServer';
+import { Client, generators, Issuer, UserinfoResponse } from 'openid-client';
+import {Sessions} from './Sessions';
+import log from 'app/server/lib/log';
+import {Permit} from './Permit';
+
+const CALLBACK_URL = '/oauth2/callback';
+const LOGIN_URL = '/oauth2/login';
+
+const LOGIN_ACTION = 'oidc-login';
+const WAIT_IN_MINUTES = 20;
+
+export class OIDCConfig {
+  private _client: Client;
+
+  public constructor(private _gristServer: GristServer) {
+  }
+
+  public async initOIDC(): Promise<void> {
+    if (!process.env.GRIST_OIDC_SP_HOST) { throw new Error('initOIDC requires GRIST_OIDC_SP_HOST to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_ISSUER) { throw new Error('initOIDC requires GRIST_OIDC_IDP_ISSUER to be set'); }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_ID) {
+      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_ID to be set');
+    }
+    if (!process.env.GRIST_OIDC_IDP_CLIENT_SECRET) {
+      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_SECRET to be set');
+    }
+    const spHost: string = process.env.GRIST_OIDC_SP_HOST;
+    const issuer = await Issuer.discover(process.env.GRIST_OIDC_IDP_ISSUER);
+    this._client = new issuer.Client({
+      client_id: process.env.GRIST_OIDC_IDP_CLIENT_ID,
+      client_secret: process.env.GRIST_OIDC_IDP_CLIENT_SECRET,
+      redirect_uris: [ new URL(CALLBACK_URL, spHost).href ],
+      response_types: ['code'],
+    });
+  }
+
+  public addEndpoints(app: express.Application, sessions: Sessions): void {
+    app.get(LOGIN_URL, this.handleLogin.bind(this));
```

Comment:
Is this endpoint (`LOGIN_URL`) needed at all? I think `/login`, `/signup`, and `/signin` might all do the same thing (by redirecting ultimately to `this.getLoginRedirectUrl()`)

Reply:
You're right! It's useless, I am removing it